src/vos/vos_layout.h

```diff
@@ -72,24 +73,70 @@ struct vos_irec_df;
 POBJ_LAYOUT_BEGIN(vos_pool_layout);
 
 POBJ_LAYOUT_ROOT(vos_pool_layout, struct vos_pool_df);
-POBJ_LAYOUT_TOID(vos_pool_layout, struct vos_cont_table_df);
 POBJ_LAYOUT_TOID(vos_pool_layout, struct vos_cont_df);
 POBJ_LAYOUT_TOID(vos_pool_layout, struct vos_dtx_table_df);
 POBJ_LAYOUT_TOID(vos_pool_layout, struct vos_dtx_entry_df);
 POBJ_LAYOUT_TOID(vos_pool_layout, struct vos_dtx_record_df);
-POBJ_LAYOUT_TOID(vos_pool_layout, struct vos_obj_table_df);
 POBJ_LAYOUT_TOID(vos_pool_layout, struct vos_obj_df);
 POBJ_LAYOUT_TOID(vos_pool_layout, struct vos_krec_df);
 POBJ_LAYOUT_TOID(vos_pool_layout, struct vos_irec_df);
+POBJ_LAYOUT_TOID(vos_pool_layout, struct vos_gc_bin_df);
+POBJ_LAYOUT_TOID(vos_pool_layout, struct vos_gc_bag_df);
+
 POBJ_LAYOUT_END(vos_pool_layout);
 
+struct vos_gc_bin_df {
+	/** address of the first(oldest) bag */
+	umem_off_t		bin_bag_first;
+	/** address of the last(newest) bag */
+	umem_off_t		bin_bag_last;
+	/** max bag size in this bin */
+	uint16_t		bin_bag_size;
+	/** total number of bags within this bin */
+	uint16_t		bin_bag_nr;
+	/**
+	 * reserved: max number of bags within this bin.
+	 * TODO: we should set a limit for number of bags per bin, and start
+	 * to eagerly run GC and free spaces if there are too many bags and
+	 * queued items.
+	 */
+	uint16_t		bin_bag_max;
+	/** reserved */
+	uint16_t		bin_pad16;
+};
 
-/**
- * VOS container table
- * PMEM container index in each pool
- */
-struct vos_cont_table_df {
-	struct btr_root		ctb_btree;
+struct vos_gc_bag_df {
+	/** index of the first item in FIFO */
+	uint16_t		bag_item_first;
+	/** index of the last item in FIFO */
+	uint16_t		bag_item_last;
+	/** number of queued items in FIFO */
+	uint16_t		bag_item_nr;
+	/** reserved */
+	uint16_t		bag_pad16;
+	/** next GC bag chained on vos_gc_bin_df */
+	umem_off_t		bag_next;
+	struct vos_gc_item {
+		/* address of the item to be freed */
+		umem_off_t		it_addr;
+		/** Reserved, argument for GC_VEA/BIO (e.g. size of extent) */
```

Comment:
Isn't he size of extent is already in the entry to be freed? (What else VEA arg needs be per item and persistently stored?) If we get rid of this, the number of items in a bag can be doubled.

Reply:
If we flatten the value tree in the future, vea_free() has to take a "blk_count" as parameter. And we definitely don't want to store bio_addr_t in bag, because it's even larger.

Reply:
I mean could we store the whole entry which includes the bio_addr_t in above 'it_addr' and do the vea free along with entry free (we could get the type of 'it_addr' from the bin type)?

Anyway, this can be re-considered or adjusted later.

Reply:
Right, let's consider this in the future, I reserved the 8-bytes just in case I forget. If it's indeed useless, we can remove it anytime we want